### PR TITLE
feat: compaction integration

### PR DESCRIPTION
**Pull request by v0y4g3r**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

- Integrate compaction logic, triggers compaction on flush when SST num in level 0 reach some threshold.
- Allow `AccessLayer` trait to accept an enum of data source so it can now accept both rows from memtable or chunk reader.
- A bunch of mock components for compaction that can be used in tests.
- Add some compaction related configs.

## Checklist

- [x]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.

## Refer to a related PR or issue link (optional)
#930

### CODE REVIEW

File: src/storage/src/compaction.rs [hunk 1]
```diff
@@ -13,9 +13,22 @@
 // limitations under the License.
 
 mod dedup_deque;
+pub mod noop;
```
Comment (evenyag) [inline src/storage/src/compaction.rs:16]: Is it only used in tests? If so, we could add a `cfg(test)` attribute.

File: src/storage/src/compaction/scheduler.rs [hunk 2]
```diff
@@ -186,15 +202,19 @@ impl<R: CompactionRequest, T: CompactionTask, P: Picker<R, T>> CompactionHandler
...
-                    while let Some((table_id,  req)) = self.poll_task().await {
+                    while let Some((region_id,  req)) = self.poll_task().await {
                         if let Ok(token) = limiter.acquire_token(&req) {
-                            debug!("Executing compaction request: {}", table_id);
-                            self.handle_compaction_request(req, token).await;
+                            debug!("Executing compaction request: {}", region_id);
+                            if let Err(e) = self.handle_compaction_request(req, token).await{
+                                error!(e; "Failed to submit compaction task for region: {}", region_id);
+                            }else{
+                                info!("Submitted region compaction task: {}", region_id);
```
Comment (evenyag) [inline src/storage/src/compaction/scheduler.rs:211]: Looks like there are some format issues

File: src/storage/src/lib.rs [hunk 3]
```diff
@@ -14,10 +14,12 @@
 
 //! Storage engine implementation.
 
+extern crate core;
```
Comment (evenyag) [inline src/storage/src/lib.rs:17]: Why do we import core?

File: src/storage/src/flush.rs [hunk 4]
```diff
@@ -160,6 +164,8 @@ pub struct FlushJob<S: LogStore> {
     pub wal: Wal<S>,
     /// Region manifest service, used to persist metadata.
     pub manifest: RegionManifest,
+    /// Callbacks that get invoked on flush success.
+    pub on_success: tokio::sync::Mutex<Option<FlushCallback>>,
```
Comment (evenyag) [inline src/storage/src/flush.rs:168]: Seems we don't need the mutex.

File: src/storage/src/region.rs [hunk 5]
```diff
@@ -107,13 +108,14 @@ impl<S: LogStore> Region for RegionImpl<S> {
 ///
 /// Contains all necessary storage related components needed by the region, such as logstore,
 /// manifest, memtable builder.
-pub struct StoreConfig<S> {
+pub struct StoreConfig<S: LogStore> {
```
Comment (evenyag) [inline src/storage/src/region.rs:111]: Could we remove this trait bound on the config struct? So we don't always need a LogStore to construct the StoreConfig.

File: src/storage/src/region/writer.rs [hunk 6]
```diff
@@ -541,6 +544,10 @@ impl WriterInner {
             return Ok(());
         }
 
+        // TODO(hl): we need to pass `compaction_after_flush` config here to determine whether to
```
Comment (evenyag) [inline src/storage/src/region/writer.rs:547]: I think it's reasonable to always try to schedule compaction here.

File: src/storage/src/region/writer.rs [hunk 7]
```diff
@@ -561,4 +569,47 @@ impl WriterInner {
...
+        let schedule_compaction_cb = Box::pin(async move {
+            let level0_file_num = shared_data
+                .version_control
+                .current()
+                .ssts()
+                .level(0)
+                .file_num();
+
+            // TODO(hl): Max file num in level 0 should be configurable.
+            if level0_file_num <= 10 {
```
Comment (evenyag) [inline src/storage/src/region/writer.rs:597]: Could you define it as a constant?

File: src/storage/src/sst/parquet.rs [hunk 8]
```diff
@@ -464,7 +445,7 @@ impl ArrowPredicate for PlainTimestampRowFilter {
     }
 
     fn evaluate(&mut self, batch: RecordBatch) -> std::result::Result<BooleanArray, ArrowError> {
-        let ts_col = batch.column(self.timestamp_index);
+        let ts_col = batch.column(0);
```
Comment (evenyag) [inline src/storage/src/sst/parquet.rs:448]: Do we provide any guarantee that the first column is timestamp index column?

File: src/storage/src/compaction.rs (hunk 1 above)
Comment (v0y4g3r) [reply src/storage/src/compaction.rs:16]: mods marked with `#cfg(test)` cannot be shared between crates. These components are used in instance tests.

File: src/storage/src/region.rs (hunk 5 above)
Comment (v0y4g3r) [reply src/storage/src/region.rs:111]: This trait bound is introduced by `CompactionRequestImpl`, `CompactionRequestImpl` has a field `Wal<S>` where `S: LogStore`.

File: src/storage/src/sst/parquet.rs (hunk 8 above)
Comment (v0y4g3r) [reply src/storage/src/sst/parquet.rs:448]: We've passed a projection with only timestamp column to build predicate, when it's evaluated only the timestamp column will be passed thru.

File: src/storage/src/region/writer.rs (hunk 7 above)
Comment (v0y4g3r) [reply src/storage/src/region/writer.rs:597]: > Could you define it as a constant?

I added an config item `max_file_in_level0`.

File: src/storage/src/flush.rs (hunk 4 above)
Comment (v0y4g3r) [reply src/storage/src/flush.rs:168]: Without the mutex the FlushJob will not be `Sync` so referencing `self` across await point won't be possible when calling `run`:
https://github.com/GreptimeTeam/greptimedb/blob/b4e7dcb74ee30147385c6ae447e82e401616395d/src/storage/src/flush.rs#L247-L255

**Review by evenyag — APPROVED** [review]
LGTM

File: src/storage/src/compaction/noop.rs [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+impl<R> Debug for NoopCompactionScheduler<R> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("NoopCompactionScheduler<...>").finish()
+    }
+}
+
+#[derive(Default, Debug)]
+pub struct NoopCompactionRequest;
+#[derive(Default, Debug)]
+pub struct NoopCompactionPicker;
```
Comment (waynexia) [inline src/storage/src/compaction/noop.rs:45]:
```suggestion
#[derive(Default, Debug)]
pub struct NoopCompactionRequest;

#[derive(Default, Debug)]
pub struct NoopCompactionPicker;
```
😛

File: src/storage/src/test_util/config_util.rs [hunk 10]
```diff
@@ -51,13 +52,15 @@ pub async fn new_store_config(
...
+    let compaction_scheduler = Arc::new(NoopCompactionScheduler::default());
     StoreConfig {
         log_store,
         sst_layer,
         manifest,
         memtable_builder: Arc::new(DefaultMemtableBuilder::default()),
         flush_scheduler,
         flush_strategy: Arc::new(SizeBasedStrategy::default()),
+        compaction_scheduler,
+        engine_config: Arc::new(Default::default()),
```
Comment (waynexia) [inline src/storage/src/test_util/config_util.rs:64]: [`Arc` derives `Default`](https://doc.rust-lang.org/std/sync/struct.Arc.html#impl-Default-for-Arc%3CT%3E), if you want to change

**Review by waynexia — APPROVED** [review]
L6TM

**Pull request merged**